### PR TITLE
Factor out memo parsing from asyncrpcoperation_sendmany

### LINKED ISSUE #6251: Extract common memo-handling for RPC
Memo handling is duplicated in z_sendmany and z_mergetoaddress, and missing from z_shieldcoinbase. We should have a common implementation that can be shared by any RPC calls that can take advantage of memos.

(Actually getting all the operations to _use_ the extracted memo can be deferred until more of the commonalities are extracted – see #2979)

### PR DESCRIPTION
**Pull request by sellout**
Fixes #6251.

### CODE REVIEW

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 1]
```diff
@@ -559,15 +560,14 @@ uint256 AsyncRPCOperation_sendmany::main_impl() {
...
+                        ovks.second, addr, r.amount,
+                        r.memo.has_value() ? r.memo.value().ToBytes() : Memo::NoMemo().ToBytes());
             },
             [&](const libzcash::OrchardRawAddress& addr) {
-                auto value = r.amount;
-                auto memo = r.memo.has_value() ? std::optional(get_memo_from_hex_string(r.memo.value())) : std::nullopt;
-                builder_.AddOrchardOutput(ovks.second, addr, value, memo);
+                builder_.AddOrchardOutput(
+                        ovks.second, addr, r.amount,
+                        r.memo.has_value() ? std::optional(r.memo.value().ToBytes()) : std::nullopt);
```
[the remark this answers is not in the document]
Comment (sellout) [reply src/wallet/asyncrpcoperation_sendmany.cpp:570]: Yeah, I’ve gotten excited by those docs a few times, only to notice the “C++23” and get sad again 😢

File: src/wallet/memo.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        return Memo();
+    }
+
+    static std::variant<MemoError, Memo> FromHex(const std::string& memoHex) {
+        Memo result;
+        std::vector<unsigned char> rawMemo = ParseHex(memoHex.c_str());
+
+        // If ParseHex comes across a non-hex char, it will stop but still return results so far.
+        size_t slen = memoHex.length();
+        if (slen % 2 != 0 || (slen > 0 && rawMemo.size() != slen / 2)) {
```
Comment (daira) [inline src/wallet/memo.h:38]: I think this is equivalent:
```suggestion
        if (slen != rawMemo.size() * 2) {
```

File: src/wallet/memo.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            return MemoError::MemoTooLong;
+        }
+
+        for (int i = 0; i < ZC_MEMO_SIZE; i++) {
+            if (i < rawMemo.size()) {
+                result.value[i] = rawMemo[i];
+            } else {
+                result.value[i] = 0x0;
+            }
+        }
```
Comment (daira) [inline src/wallet/memo.h:52]:
```suggestion
        auto rest = std::copy(rawMemo.begin(), rawMemo.end(), result.value.begin());
        std::fill(rest, result.value.end(), 0);
```

File: src/wallet/memo.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                                "Invalid parameter, expected memo data in hexadecimal format.");
+                    case MemoError::MemoTooLong:
+                        throw std::runtime_error(strprintf(
+                                "Invalid parameter, memo is longer than the maximum allowed %d characters.",
+                                ZC_MEMO_SIZE));
+                    default:
+                        assert(false);
+                }
+                // unreachable, but the compiler can't tell
+                return Memo::NoMemo();
```
Comment (daira) [inline src/wallet/memo.h:75]:
```suggestion
                }
                assert(false);
```

**Review by daira — APPROVED** [review]
utACK with nonblocking suggestions.

File: src/wallet/memo.h (hunk 2 above)
Comment (sellout) [reply src/wallet/memo.h:38]: Yeah, that’s certainly an improvement. I’m curious why the existing implementation wouldn’t error on negative even `slen` – seems like a bug. Definitely the kind of thing that should have been commented if not a bug!

**Review by daira — APPROVED** [review]
utACK

**Review by nuttycom — CHANGES_REQUESTED** [review]
One minor change to eliminate a bit more duplicated code.

**Review by nuttycom — APPROVED** [review]
utACK; the remaining changes to error handling can be done later.

**Pull request merged**